### DISCUSSION TITLE
Storing data from node contraction

### DISCUSSION THREAD

**Discussion by jamartinh** [networkx/networkx#7781] [Ideas]
Hi, I am observing that when using functions "contracted_nodes" "identified_nodes" then nodes start to have an undocumented attribute "contraction".

Will this be a constant ? a supported feature ?

**Comment by dschult** [comment]
From [the docs for contracted_nodes](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.minors.contracted_nodes.html):

> For non-multigraphs where u and v are adjacent to a third node w, the edge (v, w) will be contracted into the edge (u, w) with its attributes stored into a “contraction” attribute.

**Comment by jamartinh** [comment]
Thanks! I guess we can improve the doc. Perhaps contracting and edge into
another edge at least for me is not a common concept, although is the same
of contractinf a vertex in the  line graph. Also, what I see is that the
attribute contraction appears as a property of the node and from the doc it
is ambiguous if it should he a property of the node or the new edge  (u,w).
What do you think? Or I am missing something else?
Best,
Jose

**Reply by dschult** [reply]
Yes -- you are correct that the edge isn't "contracted", but I don't know the word to replace that. The nodes are contracted, and the edge (v, w) becomes an edge (u, w). I guess the word "contracted" vs "becomes" is more descriptive. But neither are standard terms. Also, the doc doesn't say whether the attribute is a node attribute or an edge attribute. And it should (a node attribute).
:)

**Reply by rossbar** [reply]
As I understand it, both the edges and the nodes in the resulting graph will have `"contraction"` attributes that keep any info related to contracted nodes/edges and their attributes. +1 for documenting this better!

I'm also wondering if it might be worthwhile to add an option to disable this behavior. For example, adding `store_contraction=True` (name flexible) kwarg that stores info about the contraction by default, but can be set to False to avoid adding additional attributes on the returned/modified graph.

@jamartinh would this have been a useful feature for your use-case?

**Comment by jamartinh** [comment]
Thanks, for my use case, simple graphs, the info about the contracted
edges/vertices should belong to the surviving vertex since after that the
removed vertex and edge does not exists any more until the graph is not
simple.

Another thing is that if we want to preserve  information to revert an
contraction/identification, that's another thing.

The optional store param seems useful and can work both as a graph
attribute for all contractions or/and a kwarg parameter in every call con
contract to just track or not some operations.

In any case,  this addition is useful but then it should need a more
"protagonism" in the documentation and more formal support to mature and
not be a kind of obscure feature.

Edge/vertex contraction is a very essential operation in frapg theory that
deserves a full analysis and support in graph libraries.
Best,
Jose

**Reply by rossbar** [reply]
I took a stab at improving the documentation for the current behavior in #7856 - LMK what you think @jamartinh !

**Comment by rossbar** [comment]
FWIW I think I actually did stumble on a real motivating use-case for adding the option to not store contraction info!

I'm working on an `nx-guide` based on using NetworkX to explore the [Taxonomy Database](https://www.ncbi.nlm.nih.gov/taxonomy). While exploring this data, I noticed there are many, *many* more categorical classes than the typical kingdom, phylum, order etc[^1]. One thing I'd like to try is contracting all the nodes for the "extra" categories to see if I can't condense the taxonomy down to the "standard" categories while preserving the hierarchy.

The motivating part comes in here: the `DiGraph` representing the complete taxonomy is about 1.6 GB in memory. For this particular use-case (where I only care about the relationships in the "contracted" graph, not their relationship to the original uncontracted taxonomy) it'd be great to take advantage of the memory-savings you'd get from dropping all the contraction info!

Just wanted to drop this in here - I'll consider opening a PR for making the contraction info optional (unless someone beats me to it!)

[^1]: As a concrete example - if you compute the shortest path from `"Homo sapiens"` back to the root node of the taxonomy tree, you get a path with 31 nodes in it!

**Comment by jamartinh** [comment]
I agree, an abstract representation is an abstract representation, a graph.
Otherwise the data structure is not a graph but another different data
structure.
